keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,199 @@
+---
+kep-number: 13
+title: pull based audit events stream
+authors:
+  - "@CaoShuFeng"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-06-11
+last-updated: 2018-06-11
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+
+---
+
+# Pull based audit events stream
+
+## Table of Contents
+
+* [Pull based audit events stream](#pull-based-audit-events-stream)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Make audit backends dynamicly registerable](#make-audit-backends-dynamicly-registerable)
+    * [endpoints and query parameters](#endpoints-and-query-parameters)
+    * [Bandwidth limit Configuration](#bandwidth-limit-configuration)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This proposal aims at providing a pull-based stream for audit logging. This would look similar to a watch request to the API server, and provide stream of audit events. The stream does not return any previous events (i.e. no associated storage), it would only stream events that occurred after the stream was opened. Two non resource-url endpoints are installed to provide the stream. '/audits/' for all audit events generated in the API server, and '/audits/{namespace}' for audits events in the specified namespace. And query parameters could be used to get subset of audit events.
```

Comment:
> Two non resource-url endpoints

I think these should be resource endpoints, even though they're "pseudo resources"

Reply:
When implementing pseudo resources, there are dragons I heard.

Reply:
Doubling the load for each request.

Reply:
> Doubling the load for each request.

:cry:  :cry:  :cry: 
Yes, but I haven't figure out another way to implement this in HA environment.

Reply:
I did some research about aggregation server. kube-aggregator could also aggregate audit events if the aggregated api-server has the same endpoint exposed.

But if the kube-apiserver is behind a load balancer, audit events in another kube-apiserver could not be returned, because no storage is involved.
Is that OK?

Reply:
It is a major constraint, but might be acceptable depending on the environment. It just means that your auditing system has to know and query all HA instances of kube-apiserver.